### PR TITLE
[NEW][ENTERPRISE] Support for custom Livechat registration form fields

### PR DESCRIPTION
**Pull request by renatobecker**
Related to https://github.com/RocketChat/Rocket.Chat.Livechat/pull/407

This implementation adds new properties to the `Livechat custom fields`.
These changes will impact both `Livechat Registration Form` and `Visitor Edit Panel`.

For the the Livechat Widget, only fields with the `scope = visitor`, `visibility = visible` and `public = true`, will be shared. 

The new fields supported(Enterprise edition) are described below:

- `type (text | select)`
- `required (true | false)`
- `options (used on Select components | optional)`
- `defaultValue (optional)`
- `public (true | false)`

Screen Shot 2020-05-12 at 18 53 06

On the widget side, it will look like this(Now displaying validation errors):

Screen Shot 2020-05-12 at 12 10 33

Once the Visitor is registered and starts a new conversation, the `custom fields` information will be displayed on the `Omnichannel Room Info panel` as you can see here:

Screen Shot 2020-05-08 at 17 29 04

In addition, the fields with the `type = select` will be rendered properly on the `Visitor Edit` panel.

Screen Shot 2020-05-12 at 19 00 36

### CODE REVIEW

File: ee/app/livechat-enterprise/server/hooks/onLoadConfigApi.js [hunk 1]
```diff
@@ -1,35 +1,16 @@
...
-	const [inq] = await LivechatInquiry.getCurrentSortedQueueAsync({ _id });
-
-	let queueInfo;
-	if (inq) {
-		const { position, department } = inq;
-		queueInfo = await normalizeQueueInfo({ position, department });
-	}
-
-	return { queueInfo };
+	return Object.assign({ config }, { ...options });
```
Comment (ggazzo) [inline ee/app/livechat-enterprise/server/hooks/onLoadConfigApi.js:15]: you are creating a new object to assign a another new object... there are many ways to do that... I will suggest some of them 
```javascript
     return { config , ...options };
     return Object.assign({ config }, options);

or 
return  { 
   config: { 
      ...queueInfo && { queueInfo },
      ...customFields && { customFields },
   },
   ...options,
}

File: ee/app/livechat-enterprise/server/lib/Helper.js [hunk 2]
```diff
@@ -216,3 +223,39 @@ export const updatePriorityInquiries = (priority) => {
...
+
+	const [inq] = await LivechatInquiry.getCurrentSortedQueueAsync({ _id });
+
+	let queueInfo;
+	if (inq) {
+		const { position, department } = inq;
+		queueInfo = await normalizeQueueInfo({ position, department });
+	}
+
+	return queueInfo;
```
Comment (ggazzo) [inline ee/app/livechat-enterprise/server/lib/Helper.js:260]:
```suggestion
	if (!inq) {
		return;
	}	
	return normalizeQueueInfo(inq);
```

File: app/livechat/client/views/app/tabbar/visitorEdit.js [hunk 3]
```diff
@@ -11,6 +11,16 @@ import { getCustomFormTemplate } from '../customTemplates/register';
...
 
+const getCustomFieldsByScope = (customFields = [], data = {}, filter, disabled) =>
+	customFields
+		.filter(({ visibility, scope }) => visibility !== 'hidden' && scope === filter)
+		.map(({ _id: name, scope, label, ...extraData }) => {
+			const value = data[name] ? data[name] : '';
+			return { name, label, scope, value, disabled, ...extraData };
+		});
+
+const isCustomFieldDisabled = () => hasPermission('view-livechat-room-customfields') && !hasPermission('edit-livechat-room-customfields');
```
Comment (ggazzo) [inline app/livechat/client/views/app/tabbar/visitorEdit.js:22]:
```suggestion
const isCustomFieldDisabled = () => !hasPermission('edit-livechat-room-customfields');
```

**Comment by renatobecker** [comment]
> Good feature!!! I splitted the review in two parts that one is just the first step, I tested but not deeply yet, just suggested some important improvements, focusing mainly in code style and how to use better the features that the language offers

Thanks @ggazzo for review.
I just submitted some changes that you suggested.

**Comment by mmsbr** [comment]
This Enterprise features means that it is only available on official paid server?

**Comment by renatobecker** [comment]
> This Enterprise features means that it is only available on official paid server?

It means, actually, that you need to get a `license` to run the Enterprise features.
If you take a look at our [Pricing Page](https://rocket.chat/pricing) you'll notice that some of the features are only available on the Enterprise plan.

Thanks.

**Pull request merged**